Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,125 @@
+{
+    s_connections.set(client_id, *this);
+}
+
+ClientConnection::~ClientConnection()
+{
+}
+
+void ClientConnection::die()
+{
+    s_connections.remove(client_id());
+    exit(0);
+}
+
+Messages::FileSystemAccessServer::RequestFileReadResponse ClientConnection::request_file_read(String const& path)
+{
+    auto fd = open(path.characters(), O_RDONLY);
+
+    // TODO: Prompt user via new GUI control, the program is trying to access a file by itself.
+    // TODO: if root file...
+
+    return { -1, IPC::File(fd, IPC::File::CloseAfterSending) };
+}
+
+Messages::FileSystemAccessServer::PromptFileReadResponse ClientConnection::prompt_file_read(String const& path_to_view)
+{
+    auto app = GUI::Application::construct(0, nullptr);
+    auto user_picked_file = GUI::FilePicker::get_open_filepath(nullptr, "Select file", path_to_view);
+    app->exec();
+    app->quit();
+
+    if (user_picked_file.has_value()) {
+        auto fd = open(user_picked_file.value().characters(), O_RDONLY);
+
+        if (fd < 0) {
+            dbgln(OSError(errno).string());
+            return { -1, Optional<IPC::File> {} };
+        }
+
+        return { -1, IPC::File(fd, IPC::File::CloseAfterSending) };
+    }
+
+    return { -1, Optional<IPC::File> {} };
+}
+
+Messages::FileSystemAccessServer::RequestFileWriteResponse ClientConnection::request_file_write(String const& path)
+{
+    // TODO: Prompt user via new GUI control, the program is trying to access a file by itself.
+    // TODO: if root file...
+
+    (void)path;
+    return { -1, Optional<IPC::File> {} };
+}
+
+Messages::FileSystemAccessServer::PromptFileWriteResponse ClientConnection::prompt_file_write(String const& path_to_view)
+{
+    auto app = GUI::Application::construct(0, nullptr);
+    auto user_picked_file = GUI::FilePicker::get_save_filepath(nullptr, "save", "txt", path_to_view);
+    app->exec();
+    app->quit();
+
+    if (user_picked_file.has_value()) {
+        auto fd = open(user_picked_file.value().characters(), O_WRONLY | O_CREAT, 0644);
+
+        if (fd < 0) {
+            dbgln(OSError(errno).string());
+            return { -1, Optional<IPC::File> {} };
+        }
+
+        return { -1, IPC::File(fd, IPC::File::CloseAfterSending) };
+    }
+
+    return { -1, Optional<IPC::File> {} };
+}
+
+void ClientConnection::request_download_file(URL const& url, String const& path)
+{
+    // TODO: Prompt user via new GUI control, the program is trying to access a file by itself.
+    // TODO: if root file...
+
+    (void)url;
+    (void)path;
+    TODO();
+}
+
+void ClientConnection::prompt_download_file(URL const& url, String const& filename, String const& ext)
+{
+    auto app = GUI::Application::construct(0, nullptr);
+    auto user_picked_file = GUI::FilePicker::get_save_filepath(nullptr, filename, ext, Core::StandardPaths::downloads_directory());
+    app->exec();
+    app->quit();
```

Comment:
Do you need to always instantiate a `GUI::Application` here? I assume it's only about the event loop, so maybe a `Core::EventLoop` suffices.

Reply:
This was a little workaround to remove this error:
```
FileSystemAccessServer(35:35): USERSPACE(35) ASSERTION FAILED: !s_default_font_query.is_empty()
../../Userland/Libraries/LibGfx/FontDatabase.cpp:48
[#0 FileSystemAccessServer(35:35)]: Terminating FileSystemAccessServer(35) due to signal 6
```

It feels a little dirty to do it the current way, but I think the only alternative is to create a helper application?

Reply:
The GUI::Applications clash in prompt_download if you try and download two things. I might split the DownloadWidget to a completely seperate application.

Reply:
I replaced the main `Core::EventLoop` with a single `GUI::Application` like `NotificationServer` and got rid of all the other dodgy ones.

https://github.com/SerenityOS/serenity/pull/7964/commits/d5683b61ac493685d2e7423abf137858123c2553